Fixes #5926 - hide sensitive parameter values

ares opened the pull request:
User can check to hide value when creating or editing global parameter.
The value is masked by **\* and is also not displayed when overriding.

This is meant as short-term solution. Long-term we should move to smart variables and smart class parameters and add proper password support there. I know that similar patch would be handy for smart variables and smart class parameters, but it seems too much work which would be dropped later.

ares committed:
Fixes #5926 - hide sensitive parameter values

User can check to hide value when creating or editing global parameter.
The value is masked by *** and is also not displayed when overriding.

dLobatog commented:
This makes sense to me, as I've put keys as parameters in the past, passwords, etc..

:+1: I'll review and comment later.

dLobatog commented:
Looks like a good foundation for what you describe, I think we can probably do the same we do with ComputeResource passwords.

![](https://i.imgur.com/81hpXJW.png)

My only con is the 'X' button looks a little packed between hide and remove. I personally prefer just icons and would remove the words and leave just the "X" and something like this ![](https://cdn4.iconfinder.com/data/icons/sibcode-line-simple/512/hide-24.png), but I'd wait for more feedback on that :smile:

ares commented:
Ah, it was not broken in my browser :-) I agree with icons only, if anyone can point me to right icon, I'll change it

domcleal commented:
Why aren't we doing the long-term solution of updating smart class parameters?  We're not dropping them, if anything, we're dropping this.  I don't really like adding new features to the "old" method and not the recommended implementation (smart class params).

ares commented:
To track the similar functionality for smart class params I created http://projects.theforeman.org/issues/6007

domcleal commented on app/assets/javascripts/hidden_values.js:
```diff
@@ -0,0 +1,41 @@
+function turn_text_switch() {
```
This file also needs listing in config/environments/production.rb's javascript setting.

ares replied:
Fixed, sorry it took me that long time. As a punishment I rebased this PR.

lzap commented:
Looks like jenkins was stuck. [test]

ares commented:
now it's green :-)

dLobatog commented:
Thanks for rebasing, can you fix the hide/remove thing? It looks very weird with remove in a different line. I think we agreed to use a :mag: for the hide button.

ares commented:
@eLobato the issue with using icon is that you don't see the status (whether it's on or off) also I think that magnifier does not make clear that it toggles password visibility, anyway I fixed it so both icon and text is inline by narrowing the value column. Also I fixed the hiding on override parameters.

ares commented:
[test] I think rundeck in hosts controller is random and not related

dLobatog commented:
@domcleal @ohadlevy Are we OK with this? I'm fine with merging it since we don't have any dates as far as I know to finish the migration to smart variables or smart class parameters.

ares commented:
Rebased

ohadlevy commented on app/assets/javascripts/hidden_values.js:
```diff
@@ -0,0 +1,41 @@
+      $(switch_field).change(function () {
+        var source = $(this).closest('tr').children('td.value').children()[0];
+        if (this.checked) {
+          var target = '<input class="form-control" type="password" id="' + source.id + '" name="' + source.name + '" value ="' + source.value + '"></input>'
```
is there a way to generate it via rails, and just enable / disable the input? I'm concerned that later on we change implementation in the form helpers, this js will break.

ares replied:
I don't see an easy way, two options I see:
- use ERB so we'd have hidden_values.js.erb, first pass would generate field template here and we'd have to modify some specific attributes like id, name and such but I'm not sure how assets precompilation would work and it seems hackish
- we could create a new helper for hidden values that would generate 2 fields beforehand, one of them hidden, when checkbox is clicked, they'd have to be swapped, modifying their id, name and such which gets a bit trickier because id must remain unique so 3 changes would be needed

while second seems to be feasible, I don't see big benefit since this only happens when switching hidden flag, which I believe happens once per parameter life

ohadlevy commented on app/assets/javascripts/host_edit.js:
```diff
@@ -395,9 +395,9 @@ function override_param(item){
+  var new_param = param.closest('.tab-pane').find('td.col-md-7 [id*=host_host_parameters]:visible').parent().parent().last();
   new_param.find('[id$=_name]').val(n);
-  new_param.find('[id$=_value]').val(v);
+  new_param.find('td.col-md-7 [id$=_value]').val(v == '*****' ? '' : v);
```
can we use some data attribute instead of relaying on a string?

ohadlevy commented on app/models/parameter.rb:
```diff
@@ -24,6 +24,10 @@ def self.reassign_priorities
   end
 
+  def safe_value
+    self.hidden_value? ? '*' * 5 : self.value
```
imho, pass an attribute that its safe, and then let JS add the ******

ares replied:
then the password would be in html but we don't want it there, when overriding param values we don't need original ones

ares commented:
concerns addressed or commented, please rereview

dLobatog commented:
[test] to make cops run

dLobatog commented:
Merged as https://github.com/theforeman/foreman/commit/7669211c00d8b63945a5ffcfb5c8744396b02eec for 1.7, thanks @ares !
